pkg/kubelet/cm/devicemanager/pod_devices.go

```diff
@@ -384,3 +384,23 @@ func (rdev ResourceDeviceInstances) Clone() ResourceDeviceInstances {
 	}
 	return clone
 }
+
+// Filter takes a condition set expressed as map[string]sets.String and returns a new
+// ResourceDeviceInstances with only the devices matching the condition set.
+func (rdev ResourceDeviceInstances) Filter(cond map[string]sets.String) ResourceDeviceInstances {
+	filtered := NewResourceDeviceInstances()
+	for resourceName, deviceInstances := range rdev {
+		matchIDs, ok := cond[resourceName]
+		if !ok {
+			continue
+		}
+		filtered[resourceName] = make(DeviceInstances)
+		for devID, devInstance := range deviceInstances {
+			if !matchIDs.Has(devID) {
+				continue
+			}
+			filtered[resourceName][devID] = devInstance
+		}
+	}
+	return filtered
+}
```

Comment:
I think this can be simplified to:
```
func (rdev ResourceDeviceInstances) Filter(cond map[string]sets.String) ResourceDeviceInstances {
	filtered := NewResourceDeviceInstances()
	for resourceName, filterIDs := range cond {
		if instances, exists := range rdev[resourceName]; exists {
			for instanceID, instance := range instances {
				if filterIDs.Has(instanceID) {
					filtered[resourceName][instanceID] = instance
					break
				}
			}
		}
	}
	return filtered
}
```
or if you prefer to reduce the nesting:
```
func (rdev ResourceDeviceInstances) Filter(cond map[string]sets.String) ResourceDeviceInstances {
	filtered := NewResourceDeviceInstances()
	for resourceName, filterIDs := range cond {
		if _, exists := range rdev[resourceName]; !exists {
			continue
		}
		for instanceID, instance := range rdev[resourceName] {
			if filterIDs.Has(instanceID) {
				filtered[resourceName][instanceID] = instance
				break
			}
		}
	}
	return filtered
}
```

Reply:
Yep, I mostly moved the code without further work :) Your second option looks nicer than mine, going for it.

Reply:
👍 Just noticed I missed the line for the following, so don't forget to put it back in :):
```
filtered[resourceName] = make(DeviceInstances)
```

Reply:
done! (and added a bit more tests)